Source/WebCore/accessibility/AXTextMarker.cpp

```diff
@@ -933,7 +933,41 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
 
         return resultMarker;
     }
-    // FIXME: Implement the other AXTextUnits.
+
+    // Paragraphs must be handled differently from word + sentence boundaries, as there is no paragraph break iterator.
+    // Rather, paragraph boundaries are based on rendered newlines and changes in nodes/renderers.
+    if (textUnit == AXTextUnit::Paragraph) {
+        size_t startLineIndex = currentRuns->lineID(runIndex).lineIndex;
+        unsigned offsetInStartLine = currentRuns->runLengthSumTo(startLineIndex - 1) - offset();
+
+        while (currentObject) {
+            RELEASE_ASSERT(currentRuns->size());
+            unsigned cumulativeOffset = direction == AXDirection::Next ? 0 : currentRuns->runLengthSumTo(startLineIndex);
+            for (size_t i = startLineIndex; i < currentRuns->size() && i >= 0; direction == AXDirection::Next ? i++ : i--) {
+                // If a text run starts or ends with a newline character, that indicates a paragraph boundary.
+                if (currentRuns->at(i).endsWithLineBreak() && (i != startLineIndex || (direction == AXDirection::Next && currentRuns->runLength(i) < offsetInStartLine))) {
+                    unsigned newlineOffsetConsideringDirection = direction == AXDirection::Next ? cumulativeOffset + currentRuns->runLength(i) - 1 : cumulativeOffset;
```

Comment:
`currentRuns->runLength(i) - 1` — subtracting one is safe because textruns should always be non-zero, right? I feel like you said that in a previous PR. Do you know if we have some `ASSERT` that checks this somewhere? e.g. maybe when we build the text runs. Or maybe we could put the `ASSERT` in `runLength` (if there isn't already an `ASSERT` somewhere) with a comment saying several places rely on this being non-zero when subtracting the result?

Reply:
Yeah, I confirmed in a previous patch that runLength can't be zero. I think an assert is a good idea—I'll go ahead and add that